### PR TITLE
singlevm: Add keystone support

### PR DESCRIPTION
**Pull request by bwarden**
* Replaces embedded values with variables defined near the top of
setup.sh.
* Generates configuration.yaml automatically and removes static copy
from repo.
* Adds CIAO_* variables to ~/local/demo.sh in a bit more sustainable
fashion.
* Cleans up some long lines.

Signed-off-by: Brett T. Warden <brett.t.warden@intel.com>

### CODE REVIEW

File: testutil/singlevm/setup.sh [hunk 1]
```diff
@@ -20,6 +25,49 @@ ciao_cnci_url="https://download.clearlinux.org/demos/ciao"
 fedora_cloud_image="Fedora-Cloud-Base-24-1.2.x86_64.qcow2"
 fedora_cloud_url="https://download.fedoraproject.org/pub/fedora/linux/releases/24/CloudImages/x86_64/images/Fedora-Cloud-Base-24-1.2.x86_64.qcow2"
 download=0
+conf_file="$ciao_bin"/configuration.yaml
+ciao_username=csr
```
Comment (bwarden) [inline testutil/singlevm/setup.sh]: I need to know which users/passwords should be created for testing.

**Comment by rbradford** [comment]
I get the following error:

testutil/singlevm/setup.sh: line 514: /home/rob/local/demo.sh: No such file or directory

And ciao-cli fails to upload to the image service.

**Comment by markdryan** [comment]
@bwarden To elaborate a bit more on the BAT tests issues, the BAT tests a number of ciao-cli commands,  some of which require admin priviliges to run, some of which can be executed with normal user privilidges.

When running tests that require normal priviliges the BAT tests use these env variables

* CIAO_USERNAME
* CIAO_PASSWORD

When running the admin tests 

* CIAO_ADMIN_USERNAME
* CIAO_ADMIN_PASSWORD

are used.  Actually, what happens here is that the BAT tests store the values of the CIAO_ADMIN_USERNAME and CIAO_ADMIN_PASSWORD in CIAO_USERNAME and CIAO_PASSWORD before running ciao-cli.

In the new PR CIAO_ADMIN_USERNAME is set to admin but I can't see where the admin user is created, i.e., I don't see a

os_create_user "$ciao_admin_username" "$ciao_admin_password"

Perhaps all we need to do is to add one and figure out how to let this user do a 

ciao-cli tenant list -all

**Comment by bwarden** [comment]
Reworking... will update tomorrow when it's ready to re-review.

**Comment by bwarden** [comment]
admin user is created by /usr/bin/bootstrap.sh, within docker keystone container, using the password passed in via KEYSTONE_ADMIN_PASSWORD.

**Comment by bwarden** [comment]
Did something change with our clearlinux/keystone container? Now it fails to start:

$ docker logs -f keystone
161116 00:57:20 mysqld_safe Logging to '/var/lib/mysql/6742b277282b.err'.
mysqld is alive
Traceback (most recent call last):
  File "/usr/sbin/keystone-manage", line 6, in <module>
    from keystone.cmd.manage import main
  File "/usr/lib/python2.7/site-packages/keystone/cmd/manage.py", line 32, in <module>
    from keystone.cmd import cli
  File "/usr/lib/python2.7/site-packages/keystone/cmd/cli.py", line 35, in <module>
    from keystone.federation import idp
  File "/usr/lib/python2.7/site-packages/keystone/federation/idp.py", line 23, in <module>
    from saml2 import client_base
  File "/usr/lib/python2.7/site-packages/saml2/client_base.py", line 11, in <module>
    from saml2.entity import Entity
  File "/usr/lib/python2.7/site-packages/saml2/entity.py", line 11, in <module>
    from Cryptodome.PublicKey import RSA
ImportError: No module named Cryptodome.PublicKey
[uWSGI] getting INI configuration from /usr/share/uwsgi/keystone/admin.ini
[uWSGI] getting INI configuration from /usr/share/uwsgi/keystone/public.ini
Traceback (most recent call last):
  File "/usr/sbin/keystone-manage", line 6, in <module>
    from keystone.cmd.manage import main
  File "/usr/lib/python2.7/site-packages/keystone/cmd/manage.py", line 32, in <module>
    from keystone.cmd import cli
  File "/usr/lib/python2.7/site-packages/keystone/cmd/cli.py", line 35, in <module>
    from keystone.federation import idp
  File "/usr/lib/python2.7/site-packages/keystone/federation/idp.py", line 23, in <module>
    from saml2 import client_base
  File "/usr/lib/python2.7/site-packages/saml2/client_base.py", line 11, in <module>
    from saml2.entity import Entity
  File "/usr/lib/python2.7/site-packages/saml2/entity.py", line 11, in <module>
    from Cryptodome.PublicKey import RSA
ImportError: No module named Cryptodome.PublicKey

**Comment by bwarden** [comment]
At this point, keystone is completely integrated. There shouldn't be any further substantive work required.

BAT currently fails, but there are few specific issues that I could use some input on handling:
- I should probably have setup.sh export CIAO_TENANT_NAME=admin since admin currently has roles in multiple projects. Without it, BAT fails due to ciao-cli's wanting to know which tenant to use.
- With tenant name provided, BAT times out after 10 minutes waiting on `ciao-cli workload list` with a template. I think this is due to https://github.com/01org/ciao/issues/802, which upstream indicates we need to handle internally.
- I cannot get `ciao-cli tenant list -all` to work at all -- Keystone itself is rejecting the request due to failing its authorization rules. Maybe I'm missing some role/project/service configuration. Trying to compare with a working Keystone, but it's extremely slow to respond to openstack commands.

**Comment by markdryan** [comment]
I'm not seeing any issues with tenants and BAT tests out of the box with this PR.  When I do a 

```
# ciao-cli tenant list
Projects for user csr
	UUID: 0458a3bfe0614a9699a59801e5c1da08
	Name: service
```

I only see one tenant and so the BAT tests run as expected.  Well some of them fail but not for the reason that multiple tenants are defined.

I can get all the BAT tests bar one to pass if I set CIAO_ADMIN_USERNAME and CIAO_ADMIN_PASSWORD to contain the same values as CIAO_USERNAME and CIAO_PASSWORD, but this is probably not the correct fix.   We should try to figure out why the admin user doesn't have the rights to execute

- node list -cnci
- ciao-cli node status 
- ciao-cli tenant list -all
- ciao-cli image add

We should also consider whether the csr user should be able to run these commands.  Currently, it can run the last three.  Should it be able to? @kaccardi What do you think?

**Comment by kaccardi** [comment]
@markdryan We require only that a user has the role of admin in either the service or admin project to be considered an admin. Our current manual cluster install instructions instruct the user to make csr an admin of the service project. https://clearlinux.org/documentation/ciao/cluster-setup.html#id17

**Comment by kaccardi** [comment]
just to let you know, I tested this on a fedora dev box, running on bare metal. It worked fine - but there was one sort of gotcha. Because I currently have to disable my fedora firewall in order to use singlevm, this also killed the DOCKER iptables chain. I needed to restart the docker service prior to running single vm (but after stopping firewall) and then all was well. I did not test bat, just tested with verify.

**Comment by kaccardi** [comment]
The failure to be able to list tenants seems to be due to #865

**Comment by kaccardi** [comment]
The failure to be able to do any image operations for any user other than admin or csr (i.e. privileged users) is due to #866

A workaround for singlevm is to have the default user be admin or csr - but ideally we'd not want to have our default user be a privileged user.

File: testutil/singlevm/setup.sh [hunk 2]
```diff
@@ -270,43 +381,96 @@ else
...
+fi
+openstack role add --project demo --user "$ciao_demo_username" user
+openstack service create --name glance --description "Image Service" image
+openstack endpoint create --region RegionOne image public   https://$ciao_host:9292
+openstack endpoint create --region RegionOne image internal https://$ciao_host:9292
+openstack endpoint create --region RegionOne image admin    https://$ciao_host:9292
+
+# admin should only be admin of the admin project. This role was created by the
+# keystone container's bootstrap.
+openstack role remove --project service --user admin admin
```
Comment (bwarden) [inline testutil/singlevm/setup.sh]: This takes care of admin having roles in multiple projects -- bootstrap.sh in the keystone container added admin to service, so I'll just undo that here. With this change (and #865), BAT now passes:

=== RUN   TestGetTenants
--- PASS: TestGetTenants (0.46s)
=== RUN   TestGetClusterStatus
--- PASS: TestGetClusterStatus (0.58s)
=== RUN   TestGetWorkloads
--- PASS: TestGetWorkloads (0.41s)
=== RUN   TestGetInstance
--- PASS: TestGetInstance (6.72s)
=== RUN   TestStartAllWorkloads
--- PASS: TestStartAllWorkloads (23.87s)
=== RUN   TestGetCNCIs
--- PASS: TestGetCNCIs (5.12s)
=== RUN   TestGetAllInstances
--- PASS: TestGetAllInstances (7.26s)
=== RUN   TestDeleteAllInstances
--- PASS: TestDeleteAllInstances (6.22s)
PASS
ok      github.com/01org/ciao/_release/bat/base_bat     50.840s
=== RUN   TestAddShowDelete
--- PASS: TestAddShowDelete (2.38s)
=== RUN   TestDeleteNonExisting
--- PASS: TestDeleteNonExisting (0.35s)
=== RUN   TestImageList
--- PASS: TestImageList (2.32s)
PASS
ok      github.com/01org/ciao/_release/bat/image_bat    5.052s
=== RUN   TestCreateBlockDevice
--- PASS: TestCreateBlockDevice (0.31s)
=== RUN   TestCopyBlockDevice
--- PASS: TestCopyBlockDevice (0.83s)
PASS
ok      github.com/01org/ciao/_release/bat/storage_bat  1.136s

**Comment by markdryan** [comment]
@bwarden BAT tests are passing for me as well now.  Hooray.  But I found another problem.  The volume service is not working.

```
$ ciao-cli volume list
F1124 17:15:37.954350   29031 volume.go:159] ciao-cli FATAL: Could not get volume service client [No suitable endpoint could be found in the service catalog.]
goroutine 1 [running]:
```

I tried to get it to work but my ignorance of keystone was too manifest.  Here's what I tried

```
openstack service create --name cinderv2 --description "Volume Service" volumev2
openstack endpoint create --region RegionOne volumev2 public   https://$ciao_host:8776
openstack endpoint create --region RegionOne volumev2 internal https://$ciao_host:8776
openstack endpoint create --region RegionOne volumev2 admin    https://$ciao_host:8776
```

When I run the commands manually I see an error about needing to specify an id.

**Comment by bwarden** [comment]
Added the necessary volume endpoints, and added a line to the cleanup script to remove the database storage directory for keystone so we always start with a clean slate.

**Pull request merged**